test/unit/gateways/shift4_test.rb

```diff
@@ -0,0 +1,588 @@
+    end.respond_with(successful_purchase_response)
+
+    assert response.success?
+    assert_equal success_message(response), 'Transaction successful'
+    assert_equal @refund_options[:invoice], invoice(response)
+    assert_equal @refund_options[:total], verify_amount(response)
+  end
+
+  def test_successful_void
+    void_params = { company_name: 'Spreedly', invoice: '3333333309' }
+    response = stub_comms do
+      @gateway.void(void_params)
+    end.respond_with(successful_purchase_response)
+
+    assert response.success?
+    assert_equal success_message(response), 'Transaction successful'
+    assert_equal void_params[:invoice], invoice(response)
+  end
+
+  def test_failed_purchase
+    @invalid_auth_token = '1111g66gw3ryke00'
+    @gateway.expects(:ssl_request).returns(failed_purchase_response)
+
+    response = @gateway.purchase(@amount, @invalid_auth_token, @options)
+    assert_failure response
+    assert_nil response.authorization
+  end
+
+  def test_failed_authorize
+    @gateway.expects(:ssl_request).returns(failed_authorize_response)
+
+    response = @gateway.authorize(@amount, @credit_card, @options)
+    assert_failure response
+    assert_nil response.authorization
+    assert response.test?
+  end
+
+  def test_failed_capture
+    @invalid_auth_token = '1111g66gw3ryke00'
+    @gateway.expects(:ssl_request).returns(failed_capture_response)
+
+    response = @gateway.capture(@amount, @invalid_auth_token, @options)
+    assert_failure response
+    assert_nil response.authorization
+    assert response.test?
+  end
+
+  def test_failed_refund
+    @invalid_auth_token = '1111g66gw3ryke00'
+    @gateway.expects(:ssl_request).returns(failed_refund_response)
+
+    response = @gateway.refund(@amount, @invalid_auth_token, @options)
+    assert_failure response
+    assert_nil response.authorization
+    assert response.test?
+  end
+
+  def test_failed_void
+    @gateway.expects(:ssl_request).returns(failed_void_response)
+
+    response = @gateway.void(@invoid_options)
+    assert_failure response
+    assert_nil response.authorization
+    assert response.test?
+  end
+
+  def test_support_scrub
+    assert @gateway.supports_scrubbing?
+  end
+
+  def test_scrub
+    assert @gateway.supports_scrubbing?
+    assert_equal @gateway.scrub(pre_scrubbed), post_scrubbed
+  end
+
+  private
+
+  def success_from(response)
+    response.params['result'].first['credential']['accessToken'].present?
+  end
+
+  def verify_amount(response)
+    response_result(response)['amount']['total']
+  end
+
+  def invoice(response)
+    response_result(response)['transaction']['invoice']
+  end
+
+  def response_result(response)
+    response.params['result'][0]
+  end
+
+  def success?(response)
+    response_result(response).present?
+  end
+
+  def token?(response)
+    response_result(response)['card']['token']['value'].present?
+  end
+
+  def success_message(response)
+    response.message
+  end
```

Comment:
Same recommendation as the remote test. Remove this and retrieve the values in-line with the assertions.